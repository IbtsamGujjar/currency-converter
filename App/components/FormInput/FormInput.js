import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { View, Text, TextInput, StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  formControl: {
    flex: 1,
  },
  label: {
    fontFamily: 'open-sans-bold',
    marginVertical: 8,
  },
  input: {
    paddingHorizontal: 2,
    paddingVertical: 5,
    borderBottomColor: '#ccc',
    borderBottomWidth: 1,
  },
  errorContainer: {
    marginVertical: 5,
  },
  errorText: {
    fontFamily: 'open-sans',
    color: 'red',
    fontSize: 13,
  },
});

const FormInput = ({
  label,
  keyboardType,
  secureTextEntry,
  onChangeText,
  initialValue,
  required,
  email,
  autoCapitalize,
  errorText,
  minLength,
  initiallyValid,
  id,
}) => {
  const [inputState, setInputState] = useState({
    value: initialValue,
    isValid: initiallyValid,
    touched: false,
  });

  useEffect(() => {
    if (inputState.touched) {
      onChangeText(id, inputState.value, inputState.isValid);
    }
  }, [inputState]);

  const handleInputChange = (text) => {
    // eslint-disable-next-line no-useless-escape
    const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    let isValid = true;
    if (required && text.trim().length === 0) {
      isValid = false;
    }
    if (email && !emailRegex.test(text.toLowerCase())) {
      isValid = false;
    }
    if (minLength != null && text.length < minLength) {
      isValid = false;
    }
    setInputState({ ...inputState, value: text, isValid });
  };
  const lostFocusHandler = () => {
    setInputState({ ...inputState, touched: true });
  };
  return (
    <View style={styles.formControl}>
      <Text style={styles.label}>{label}</Text>
      <TextInput
        style={styles.input}
        value={inputState.value}
        autoCapitalize={autoCapitalize}
        keyboardType={keyboardType}
        secureTextEntry={secureTextEntry}
        onChangeText={handleInputChange}
        onBlur={lostFocusHandler}
      />
      {!inputState.isValid && inputState.touched && (
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>{errorText}</Text>
        </View>
      )}
    </View>
  );
};

FormInput.defaultProps = {
  label: '',
  initialValue: '',
  required: true,
  email: false,
  autoCapitalize: 'none',
  errorText: '',
  minLength: 5,
  initiallyValid: false,
  id: '',
  keyboardType: 'default',
  secureTextEntry: false,
  onChangeText: () => {},
};

FormInput.propTypes = {
  required: PropTypes.bool,
  email: PropTypes.bool,
  autoCapitalize: PropTypes.string,
  errorText: PropTypes.string,
  minLength: PropTypes.number,
  initiallyValid: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  initialValue: PropTypes.string,
  keyboardType: PropTypes.string,
  secureTextEntry: PropTypes.bool,
  onChangeText: PropTypes.func,
};

export { FormInput };
