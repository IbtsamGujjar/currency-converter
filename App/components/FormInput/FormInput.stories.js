import React from 'react';
import { storiesOf } from '@storybook/react-native';

import { FormInput } from './FormInput';
import { WithFonts } from '../../../storybook/stories/decorator';

storiesOf('FormInput', module)
  .addDecorator((getStory) => <WithFonts>{getStory()}</WithFonts>)
  .add('Email', () => (
    <FormInput
      label="E-Mail"
      initialValue="abc@mail.com"
      keyboardType="email-address"
    />
  ))
  .add('Password', () => (
    <FormInput
      label="Password"
      initialValue="abc@mail.com"
      secureTextEntry
      keyboardType="default"
      required
      minLength={5}
    />
  ));
