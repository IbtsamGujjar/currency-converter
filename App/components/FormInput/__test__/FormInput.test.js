import React from 'react';
import renderer from 'react-test-renderer';
import { FormInput } from '../FormInput';

test('FormInput Snapshot', () => {
  const snapshot = renderer.create(<FormInput />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
