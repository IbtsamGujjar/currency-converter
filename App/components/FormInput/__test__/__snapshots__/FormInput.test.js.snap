// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "fontFamily": "open-sans-bold",
        "marginVertical": 8,
      }
    }
  >
    
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    keyboardType="default"
    onBlur={[Function]}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    secureTextEntry={false}
    style={
      Object {
        "borderBottomColor": "#ccc",
        "borderBottomWidth": 1,
        "paddingHorizontal": 2,
        "paddingVertical": 5,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
